Tutorial
=====================

This page describes the tutorial how to get the results from FASTQ files. 
While this tutorial uses human data, the sample scripts for human, mouse and `S. cerevisiae` are also available at `Chrros GitHub site <https://github.com/rnakato/Churros/tree/main/tutorial>`_.

.. note::

   | This tutorial assumes using the **Churros** singularity image (``churros.sif``). Please add ``singularity exec churros.sif`` before each command below.
   | Example: ``singularity exec churros.sif download_genomedata.sh``


Get data
------------------------

Here we use five histone modification data of HepG2 cells from `ENCODE Project <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE29611>`_.

.. code-block:: bash

    mkdir -p fastq
    for id in SRR227447 SRR227448 SRR227552 SRR227553 SRR227563 SRR227564 SRR227575 SRR227576 SRR227598 SRR227599 SRR227639 SRR227640
    do
        fastq-dump --gzip $id -O fastq
    done

| Then download and generate the reference dataset including genome, gene annotation and index files. **Chuross** contains scripts for that: ``download_genomedata.sh`` and ``build-index.sh``. 
| Here we specify ``GRCh38`` for genome build. See :doc:`Appendix` for the detail of genome build.

.. code-block:: bash

    mkdir -p log
    build=GRCh38
    ncore=24
    # download the genome
    download_genomedata.sh $build Ensembl-$build/ 2>&1 | tee log/Ensembl-$build
    # make Bowtie2 index
    build-index.sh -p $ncore bowtie2 Ensembl-$build
    

Prepare sample list
-------------------------------------

Churros takes two input files, ``samplelist.txt`` and ``samplepairlist.txt`` that describes the detail of samples to be analyzed.

samplelist.txt
++++++++++++++++++++++++++

``samplelist.txt`` is a **tab-delimited** file (TSV) that describes the sample labels and corresponding fastq files. 
Multiple fastq files can be specified by separateing with ``,``. 

.. code-block:: bash

    HepG2_H2A.Z     fastq/SRR227639.fastq.gz,fastq/SRR227640.fastq.gz
    HepG2_H3K4me3   fastq/SRR227563.fastq.gz,fastq/SRR227564.fastq.gz
    HepG2_H3K27ac   fastq/SRR227575.fastq.gz,fastq/SRR227576.fastq.gz
    HepG2_H3K27me3  fastq/SRR227598.fastq.gz,fastq/SRR227599.fastq.gz
    HepG2_H3K36me3  fastq/SRR227447.fastq.gz,fastq/SRR227448.fastq.gz
    HepG2_Control   fastq/SRR227552.fastq.gz,fastq/SRR227553.fastq.gz


samplepairlist.txt
++++++++++++++++++++++++++

``samplelist.txt`` is a **comma-delimited** file (CSV) that describes the ChIP/Input pairs as follows:

- ChIP-sample label
- Input-sample label
- prefix
- peak mode

ChIP and input sample labels should be identical to those in ``samplelist.txt``.
Input samples can be omitted if unavailable.
``prefix`` is used for the output files.
``peak mode`` is either ``[sharp|broad|sharp-nomodel|broad-nomodel]``. This parameter is used for peak calling by `MACS2 <https://github.com/macs3-project/MACS>`_.

.. code-block:: bash

    HepG2_H2A.Z,HepG2_Control,HepG2_H2A.Z,sharp
    HepG2_H3K4me3,HepG2_Control,HepG2_H3K4me3,sharp
    HepG2_H3K27ac,HepG2_Control,HepG2_H3K27ac,sharp
    HepG2_H3K27me3,HepG2_Control,HepG2_H3K27me3,broad
    HepG2_H3K36me3,HepG2_Control,HepG2_H3K36me3,broad


Running Churros
------------------------------------------------

``churros`` command executes all steps from mapping reads to visualization. 

.. code-block:: bash

    churros -p 12 samplelist.txt samplepairlist.txt hg38 Ensembl-GRCh38/

``-p 12`` specifies the number of CPUs. ``hg38`` is the UCSC genome build and ``Ensembl-GRCh38`` is the directory generated by ``download_genomedata.sh`` and ``build-index.sh``.

To consider genome mappability, supply ``--mpbl`` option. The mappability affects the quality check results and the read-distribution normalization in DROMPA+ but does not affect peak calling by MACS2.

.. code-block:: bash

    churros -p 12 --mpbl samplelist.txt samplepairlist.txt hg38 Ensembl-GRCh38/


The results are output in ``Churros_result/``. If you want to specify the name of output directory, use ``-D`` option.

.. code-block:: bash

    churros -p 12 --mpbl -D outputdir samplelist.txt samplepairlist.txt hg38 Ensembl-GRCh38/

``--outputpvalue`` option output the bedGraph file for -log10(p-value) of ChIP/Input enrichment.

.. code-block:: bash

    churros -p 12 --mpbl --outputpvalue samplelist.txt samplepairlist.txt $build $Ddir


The detail and output are described below.

churros_mapping: mapping reads
--------------------------------------------------

``churros_mapping`` takes FASTQ and map reads to the genome specified by Bowtie2 in default.
The mapped reads are then quality-checked and converted to BigWig files.

.. code-block:: bash

    build=hg38
    Ddir=Ensembl-GRCh38
    churros_mapping exec fastq/SRR227447.fastq.gz,fastq/SRR227448.fastq.gz HepG2_H3K36me3 $build $Ddir
    churros_mapping exec fastq/SRR227552.fastq.gz,fastq/SRR227553.fastq.gz HepG2_Control  $build $Ddir
    churros_mapping exec fastq/SRR227563.fastq.gz,fastq/SRR227564.fastq.gz HepG2_H3K4me3  $build $Ddir
    churros_mapping exec fastq/SRR227575.fastq.gz,fastq/SRR227576.fastq.gz HepG2_H3K27ac  $build $Ddir
    churros_mapping exec fastq/SRR227598.fastq.gz,fastq/SRR227599.fastq.gz HepG2_H3K27me3 $build $Ddir
    churros_mapping exec fastq/SRR227639.fastq.gz,fastq/SRR227640.fastq.gz HepG2_H2A.Z    $build $Ddir

Or you can use the ``samplelist.txt`` as follows:

.. code-block:: bash

    while read LINE; do
        LINE=($LINE)
        prefix=${LINE[0]}
        fq1=${LINE[1]}
        churros_mapping -p 12 exec $fq1 $prefix hg38 Ensembl-GRCh38
    done < samplelist.txt


Call peaks by MACS2
--------------------------------------------------

``churros_callpeak.sh`` calls peaks of the samples specified in ``samplepairlist.txt``.
Input samples can be omitted.

.. code-block:: bash

    build=hg38
    churros_callpeak samplepairlist.txt $build

    
Visualize read distributions by DROMPA+
--------------------------------------------------

``churros_callpeak.sh`` calls peaks of the samples specified in ``samplepairlist.txt``.
Input samples can be omitted.

.. code-block:: bash

    build=hg38
    Ddir=Ensembl-GRCh38
    
    mkdir -p pdf
    churros_visualize samplepairlist.txt pdf/drompa+ $build $Ddir
    churros_visualize macs/samplepairlist.txt pdf/drompa+.macspeak $build $Ddir
    churros_visualize -b 5000 -l 8000 -P "--scale_tag 100" samplepairlist.txt pdf/drompa+.bin5M $build $Ddir
    churros_visualize -p -b 5000 -l 8000 samplepairlist.txt pdf/drompa+.pval.bin5M $build $Ddir
    churros_visualize -G macs/samplepairlist.txt pdf/drompa+ $build $Ddir