Tutorial (human)
=====================

This page describes the tutorial how to get the results from FASTQ files.
While this tutorial uses human data, the sample scripts for human, mouse and `S. cerevisiae` are also available at `Chrros GitHub site <https://github.com/rnakato/Churros/tree/main/tutorial>`_.

.. note::

   | This tutorial assumes using the **Churros** singularity image (``churros.sif``). Please add ``singularity exec churros.sif`` before each command below.
   | Example: ``singularity exec churros.sif download_genomedata.sh``


Get data
------------------------

Here we use five histone modification data of HepG2 cells from `ENCODE Project <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE29611>`_.

.. code-block:: bash

    mkdir -p fastq
    for id in SRR227447 SRR227448 SRR227552 SRR227553 SRR227563 SRR227564 SRR227575 SRR227576 SRR227598 SRR227599 SRR227639 SRR227640
    do
        fastq-dump --gzip $id -O fastq
    done

| Then download and generate the reference dataset including genome, gene annotation and index files. **Chuross** contains scripts for that: ``download_genomedata.sh`` and ``build-index.sh``.
| Here we specify ``GRCh38`` for genome build. See :doc:`Appendix` for the detail of genome build.

.. code-block:: bash

    mkdir -p log
    build=hg38      # genome build
    Ddir=Referencedata_$build   # output directory
    ncore=12    # number of CPUs
    # download the genome
    download_genomedata.sh $build $Ddir 2>&1 | tee log/$Ddir
    # make Bowtie2 index
    build-index.sh -p $ncore bowtie2 $Ddir

Prepare sample list
-------------------------------------

Churros takes two input files, ``samplelist.txt`` and ``samplepairlist.txt`` that describes the detail of samples to be analyzed.

samplelist.txt
++++++++++++++++++++++++++

``samplelist.txt`` is a **tab-delimited** file (TSV) that describes the sample labels and corresponding fastq files.
Multiple fastq files can be specified by separateing with ``,``.

.. code-block:: bash

    HepG2_H2A.Z     fastq/SRR227639.fastq.gz,fastq/SRR227640.fastq.gz
    HepG2_H3K4me3   fastq/SRR227563.fastq.gz,fastq/SRR227564.fastq.gz
    HepG2_H3K27ac   fastq/SRR227575.fastq.gz,fastq/SRR227576.fastq.gz
    HepG2_H3K27me3  fastq/SRR227598.fastq.gz,fastq/SRR227599.fastq.gz
    HepG2_H3K36me3  fastq/SRR227447.fastq.gz,fastq/SRR227448.fastq.gz
    HepG2_Control   fastq/SRR227552.fastq.gz,fastq/SRR227553.fastq.gz


samplepairlist.txt
++++++++++++++++++++++++++

``samplepairlist.txt`` is a **comma-delimited** file (CSV) that describes the ChIP/Input pairs as follows:

- ChIP-sample label
- Input-sample label
- prefix
- peak mode

ChIP and input sample labels should be identical to those in ``samplelist.txt``.
Input samples can be omitted if unavailable.
``prefix`` is used for the output files.
``peak mode`` is either ``[sharp|broad|sharp-nomodel|broad-nomodel]``. This parameter is used for peak calling by `MACS2 <https://github.com/macs3-project/MACS>`_.

.. code-block:: bash

    HepG2_H2A.Z,HepG2_Control,HepG2_H2A.Z,sharp
    HepG2_H3K4me3,HepG2_Control,HepG2_H3K4me3,sharp
    HepG2_H3K27ac,HepG2_Control,HepG2_H3K27ac,sharp
    HepG2_H3K27me3,HepG2_Control,HepG2_H3K27me3,broad
    HepG2_H3K36me3,HepG2_Control,HepG2_H3K36me3,broad


Running Churros
------------------------------------------------

``churros`` command executes all steps from mapping reads to visualization.

.. code-block:: bash

    churros -p 12 samplelist.txt samplepairlist.txt hg38 Referencedata_hg38

``-p 12`` specifies the number of CPUs. ``hg38`` is the UCSC genome build and ``Referencedata_hg38`` is the directory generated by ``download_genomedata.sh`` and ``build-index.sh``.


.. .. note::
.. 
..    There are two types of genome build, Ensembl and UCSC ('GRCh38' and 'hg38' in this case). Specify UCSC build to **Churros**.

.. code-block:: bash

    churros -p 12 samplelist.txt samplepairlist.txt hg38 Referencedata_hg38

The results are output in ``Churros_result/``. If you want to specify the name of output directory, use ``-D`` option.

.. code-block:: bash

    churros -p 12 -D outputdir samplelist.txt samplepairlist.txt hg38 Referencedata_hg38

By supplying ``--comparative`` option, ``churros`` implements all-by-all sample comparisons and make correlation heatmaps.

.. code-block:: bash

    churros -p 12 --comparative samplelist.txt samplepairlist.txt $build $Ddir

``--outputpvalue`` option outputs the bedGraph file for -log10(p-value) of ChIP/Input enrichment.

.. code-block:: bash

    churros -p 12 --outputpvalue samplelist.txt samplepairlist.txt $build $Ddir

Churros consider genome mappability in default. 
The mappability affects the quality check results and the read-distribution normalization in DROMPA+ but does not affect peak calling by MACS2. 
If you want not to consider it, supply ``--nompbl`` option.

.. code-block:: bash

    churros -p 12 --nompbl samplelist.txt samplepairlist.txt $build $Ddir


The detail and output are described below.


churros_mapping: mapping reads
--------------------------------------------------

``churros_mapping`` takes FASTQ and map reads to the genome specified by Bowtie2 in default.
The mapped reads are then quality-checked and converted to BigWig files.

.. code-block:: bash

    build=hg38
    Ddir=Referencedata_hg38

    # mapping
    $sing churros_mapping -p 12 exec samplelist.txt $build $Ddir
    # output QC stats
    $sing churros_mapping header > churros.QCstats.tsv
    $sing churros_mapping stats samplelist.txt $build $Ddir >> churros.QCstats.tsv

- Output
    - bam/    ... map files (BAM format in default) and index files
    - sspout/ ... output of SSP (strand-shift profile) for quality check
    - parse2wigdir+/ ... bigWig files (100-bp, 5-kbp and 100-lbp bins in default) by parse2wig+
    - log/ ... log files


churros_callpeak: call peaks by MACS2
--------------------------------------------------

``churros_callpeak`` calls peaks of the samples specified in ``samplepairlist.txt``.
If input samples are omitted, peaks are called using ChIP samples only.

.. code-block:: bash

    churros_callpeak -p 8 samplepairlist.txt hg38

``churros_callpeak`` also outputs the correlation scores (Simpson index) and heatmaps.

- Output
    - macs/ ... peak files called by MACS2. The log files are stored in *log. ``samplepairlist.txt`` in ``macs/`` directory includes the filename of peak files that is used in ``churros_visualize``.


churros_visualize: visualize read distributions by DROMPA+
--------------------------------------------------------------------

``churros_visualize`` visualizes read distribution as pdf format.

.. code-block:: bash

    churros_visualize samplepairlist.txt drompa+ hg38 Referencedata_hg38

To specify binsize 5-kbp, supply ``-b 5000``. ``-l 8000`` means the line size for each page is 8-Mbp. ``-P "--scale_tag 100"`` indicates the scale of y-axis is 100.

.. code-block:: bash

    churros_visualize -b 5000 -l 8000 -P "--scale_tag 100" samplepairlist.txt drompa+.bin5M hg38 Referencedata_hg38

To visualize genome-wide view, supply ``-G`` option.

.. code-block:: bash

    churros_visualize -G samplepairlist.txt drompa+ hg38 Referencedata_hg38

- Output
    - pdf/ ... the pdf files and corresponding peak lists.

Highlight peak regions
+++++++++++++++++++++++++++++++++

| ``churros_visualize`` can highlight peak regions if the peak file is specified in ``samplepairlist.txt``.
| (i.e., the column of ``samplepairlist.txt`` for ``churros_visualize`` is ``<ChIP-sample>,<Input-sample>,<prefix>,<peakfile>``).
| Because ``churros_callpeak`` generated ``Churros_result/$build/macs/samplepairlist.txt`` that includes the peak files, ``churros_visualize`` highlights the peak regions by the command below:

.. code-block:: bash

    churros_visualize Churros_result/hg38/macs/samplepairlist.txt drompa+.macspeak hg38 Referencedata_hg38

Visualize p-value distribution
+++++++++++++++++++++++++++++++++++++++

Supply ``--pvalue`` option to visualize -log10(p) distribution of ChIP/input enrichment, which is recommended by `ROADMAP project <https://www.nature.com/articles/nature14248>`_ to distinguish the signal from the noise.

.. code-block:: bash

    churros_visualize --pvalue -b 5000 -l 8000 samplepairlist.txt drompa+.pval.bin5M hg38 Referencedata_hg38


(Optional) modify parameter sets for visualization manually
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

``churros_visualize`` also outputs a log file of pdf files generation
(e.g., ```Churros_result/$build/log/pdf/drompa+.PCSHARP.100.log`` for ``Churros_result/$build/pdf/drompa+.PCSHARP.100.*.pdf``).
This log file contains the command of DROMPA+ to make the pdf file at the top.

.. code-block:: bash

    head -n1 Churros_result/$build/log/pdf/drompa+.PCSHARP.100.log
    drompa+ PC_SHARP  --ls 1000 -g Referencedata_hg38/gtf_chrUCSC/chr.gene.refFlat --gt Referencedata_hg38/genometable.txt --callpeak --showchr   -i Churros_result/parse2wigdir+/HepG2_H2A.Z-bowtie2-hg38-raw-mpbl-GR.100.bw,Churros_result/parse2wigdir+/HepG2_Control-bowtie2-hg38-raw-mpbl-GR.100.bw,HepG2_H2A.Z, -i Churros_result/parse2wigdir+/HepG2_H3K4me3-bowtie2-hg38-raw-mpbl-GR.100.bw,Churros_result/parse2wigdir+/HepG2_Control-bowtie2-hg38-raw-mpbl-GR.100.bw,HepG2_H3K4me3, -i Churros_result/parse2wigdir+/HepG2_H3K27ac-bowtie2-hg38-raw-mpbl-GR.100.bw,Churros_result/parse2wigdir+/HepG2_Control-bowtie2-hg38-raw-mpbl-GR.100.bw,HepG2_H3K27ac, -i Churros_result/parse2wigdir+/HepG2_H3K27me3-bowtie2-hg38-raw-mpbl-GR.100.bw,Churros_result/parse2wigdir+/HepG2_Control-bowtie2-hg38-raw-mpbl-GR.100.bw,HepG2_H3K27me3, -i Churros_result/parse2wigdir+/HepG2_H3K36me3-bowtie2-hg38-raw-mpbl-GR.100.bw,Churros_result/parse2wigdir+/HepG2_Control-bowtie2-hg38-raw-mpbl-GR.100.bw,HepG2_H3K36me3, -o Churros_result/pdf/drompa+.PCSHARP.100 | tee -a Churros_result/pdf/drompa+.PCSHARP.100.log

Therefore, you can modify the resulting pdf files by directly modifying this command and ``-o`` option that specifies the output name.
For example, if you want to change the y-axis scale to 50, add ``--scale_tag 50`` and execute:

.. code-block:: bash

    drompa+ PC_SHARP --scale_tag 50 --ls 1000 (...) -o Churros_result/pdf/drompa+.PCSHARP.100.modified

See `DROMPAplus manual <https://drompaplus.readthedocs.io/en/latest/index.html>`_ for the detailed usage of DROMPA+.


churros_compare: compare peaks among ChIP samples
--------------------------------------------------------------------

``churros_compare`` output the heatmap of correlation of peaks among ChIP samples.
The results will be output in ``comparsion/`` directory. 
``churros_compare`` estimates peak overlap for 'all peaks' and 'top-ranked 2000 peaks'.

.. code-block:: bash

    churros_compare samplelist.txt samplepairlist.txt hg38

- Output
    - bigwigCorrelation/ ... Spearman correlation of read distributon in 100-bp and 100-kbp bins by `deepTools plotCorrelation <https://deeptools.readthedocs.io/en/develop/content/tools/plotCorrelation.html>`_ and jaccard index of peak overlap by `bedtools jaccard <https://bedtools.readthedocs.io/en/latest/content/tools/jaccard.html>`_ (for all peaks and top-ranked 2000 peaks)
    - Peak_BPlevel_overlap/ ... results of base-pair level overlap of peaks (Jaccard index) by BEDtools jaccard
    - Peak_Number_overlap/ ... results of peak-number level comparion (Simpson index). ``PairwiseComparison/`` contains the results of all pairs (overlapped peak list and Venn diagram) and the ``Peaks`` contains top-ranked peaks of samples.


.. note::

   Unlike the peak comparison implemented in ``churros_callpeak``, ``churros_compare`` evaluates the similarity of whole genome including non-peak regions. Therefore the results may reflect the genome-wide features (e.g., GC bias and copy number variations) rather than peak overlap.


churros_genPvalwig: generate P-value distribution as bedGraph
--------------------------------------------------------------------

``churros_genPvalwig`` generates -log10(P-value) distribution in bedGraph format. The P-value of upregulation and downregulation are output separately.

.. code-block:: bash

    Ddir=Referencedata_hg38
    gt=$Ddir/genometable.txt
    churros_genPvalwig samplepairlist.txt drompa+.pval hg38 $gt

- Output
    - drompa+.pval/ ... bedgraph files of p-values
